File: books/practical-bioinformatics/days/day-15/scripts/visualization.py
#!/usr/bin/env python3
"""Day 15: Publication-Quality Visualization — Python equivalent."""

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import matplotlib
import os

matplotlib.use("Agg")  # Non-interactive backend for SVG output

os.makedirs("figures", exist_ok=True)

# ── Load data ──────────────────────────────────────────────────────

de = pd.read_csv("data/de_results.csv")
print(f"Loaded DE results: {len(de)} genes")

gwas = pd.read_csv("data/gwas_results.csv")
print(f"Loaded GWAS results: {len(gwas)} variants")

# ── Scatter plot ───────────────────────────────────────────────────

x = [1.0, 2.0, 3.0, 4.0, 5.0]
y = [2.1, 3.9, 6.2, 7.8, 10.1]

fig, ax = plt.subplots(figsize=(6, 4))
ax.scatter(x, y, color="steelblue")
ax.set_title("Gene Expression Correlation")
ax.set_xlabel("x")
ax.set_ylabel("y")
plt.tight_layout()
plt.savefig("figures/scatter.svg", format="svg")
plt.close()
print("Saved figures/scatter.svg")

# ── Histogram ──────────────────────────────────────────────────────

values = [2.1, 3.5, 4.2, 5.8, 6.1, 7.3, 3.8, 5.5, 4.9, 6.7, 3.2, 5.1]

fig, ax = plt.subplots(figsize=(6, 4))
ax.hist(values, bins=6, color="steelblue", edgecolor="white")
ax.set_title("Expression Distribution")
ax.set_xlabel("Expression")
ax.set_ylabel("Count")
plt.tight_layout()
plt.savefig("figures/histogram.svg", format="svg")
plt.close()
print("Saved figures/histogram.svg")

# ── Bar chart ──────────────────────────────────────────────────────

categories = ["SNP", "Insertion", "Deletion", "MNV"]
counts = [3500, 450, 520, 30]

fig, ax = plt.subplots(figsize=(6, 4))
ax.bar(categories, counts, color="steelblue")
ax.set_title("Variant Types")
ax.set_ylabel("Count")
plt.tight_layout()
plt.savefig("figures/bar_chart.svg", format="svg")
plt.close()
print("Saved figures/bar_chart.svg")

# ── Boxplot ────────────────────────────────────────────────────────

control = [5.2, 4.8, 5.1, 4.9, 5.3, 5.0]
treated = [8.1, 7.9, 8.5, 7.6, 8.3, 8.0]
resistant = [5.5, 5.3, 5.8, 5.1, 5.6, 5.4]

fig, ax = plt.subplots(figsize=(6, 4))
ax.boxplot([control, treated, resistant], labels=["control", "treated", "resistant"])
ax.set_title("Expression by Group")
ax.set_ylabel("Expression")
plt.tight_layout()
plt.savefig("figures/boxplot.svg", format="svg")
plt.close()
print("Saved figures/boxplot.svg")

# ── Volcano plot ───────────────────────────────────────────────────

de["-log10p"] = -np.log10(de["padj"])

fig, ax = plt.subplots(figsize=(8, 6))
colors = []
for _, row in de.iterrows():
    if row["padj"] < 0.05 and row["log2fc"] > 1.0:
        colors.append("red")
    elif row["padj"] < 0.05 and row["log2fc"] < -1.0:
        colors.append("blue")
    else:
        colors.append("grey")

ax.scatter(de["log2fc"], de["-log10p"], c=colors, alpha=0.7)
ax.axhline(-np.log10(0.05), ls="--", color="grey", alpha=0.5)
ax.axvline(1.0, ls="--", color="grey", alpha=0.5)
ax.axvline(-1.0, ls="--", color="grey", alpha=0.5)
ax.set_xlabel("log2 Fold Change")
ax.set_ylabel("-log10(adjusted p-value)")
ax.set_title("A) Differential Expression")
plt.tight_layout()
plt.savefig("figures/fig1_volcano.svg", format="svg")
plt.close()
print("Saved figures/fig1_volcano.svg")

# ── MA plot ────────────────────────────────────────────────────────

fig, ax = plt.subplots(figsize=(8, 6))
colors_ma = ["red" if p < 0.05 else "grey" for p in de["padj"]]
ax.scatter(np.log10(de["baseMean"]), de["log2fc"], c=colors_ma, alpha=0.7)
ax.axhline(0, ls="--", color="grey", alpha=0.5)
ax.set_xlabel("log10(baseMean)")
ax.set_ylabel("log2 Fold Change")
ax.set_title("B) MA Plot")
plt.tight_layout()
plt.savefig("figures/fig2_ma.svg", format="svg")
plt.close()
print("Saved figures/fig2_ma.svg")

# ── Summary ────────────────────────────────────────────────────────

up = len(de[(de["padj"] < 0.05) & (de["log2fc"] > 1.0)])
down = len(de[(de["padj"] < 0.05) & (de["log2fc"] < -1.0)])
ns = len(de) - up - down

print(f"\nUpregulated (padj<0.05, log2FC>1):   {up}")
print(f"Downregulated (padj<0.05, log2FC<-1): {down}")
print(f"Not significant:                      {ns}")
print("\nDone.")

File: books/practical-bioinformatics/days/day-15/scripts/visualization.R
#!/usr/bin/env Rscript
# Day 15: Publication-Quality Visualization — R equivalent

library(ggplot2)

dir.create("figures", showWarnings = FALSE)

# ── Load data ──────────────────────────────────────────────────────

de <- read.csv("data/de_results.csv")
cat(sprintf("Loaded DE results: %d genes\n", nrow(de)))

gwas <- read.csv("data/gwas_results.csv")
cat(sprintf("Loaded GWAS results: %d variants\n", nrow(gwas)))

# ── Scatter plot ───────────────────────────────────────────────────

scatter_df <- data.frame(x = c(1, 2, 3, 4, 5), y = c(2.1, 3.9, 6.2, 7.8, 10.1))
p <- ggplot(scatter_df, aes(x = x, y = y)) +
  geom_point(color = "steelblue", size = 3) +
  labs(title = "Gene Expression Correlation") +
  theme_minimal()
ggsave("figures/scatter.svg", p, width = 6, height = 4)
cat("Saved figures/scatter.svg\n")

# ── Histogram ──────────────────────────────────────────────────────

values <- c(2.1, 3.5, 4.2, 5.8, 6.1, 7.3, 3.8, 5.5, 4.9, 6.7, 3.2, 5.1)
p <- ggplot(data.frame(x = values), aes(x = x)) +
  geom_histogram(bins = 6, fill = "steelblue", color = "white") +
  labs(title = "Expression Distribution", x = "Expression", y = "Count") +
  theme_minimal()
ggsave("figures/histogram.svg", p, width = 6, height = 4)
cat("Saved figures/histogram.svg\n")

# ── Bar chart ──────────────────────────────────────────────────────

var_df <- data.frame(
  category = c("SNP", "Insertion", "Deletion", "MNV"),
  count = c(3500, 450, 520, 30)
)
p <- ggplot(var_df, aes(x = reorder(category, -count), y = count)) +
  geom_col(fill = "steelblue") +
  labs(title = "Variant Types", x = "", y = "Count") +
  theme_minimal()
ggsave("figures/bar_chart.svg", p, width = 6, height = 4)
cat("Saved figures/bar_chart.svg\n")

# ── Boxplot ────────────────────────────────────────────────────────

box_df <- data.frame(
  group = rep(c("control", "treated", "resistant"), each = 6),
  expression = c(5.2, 4.8, 5.1, 4.9, 5.3, 5.0,
                 8.1, 7.9, 8.5, 7.6, 8.3, 8.0,
                 5.5, 5.3, 5.8, 5.1, 5.6, 5.4)
)
p <- ggplot(box_df, aes(x = group, y = expression, fill = group)) +
  geom_boxplot() +
  labs(title = "Expression by Group", y = "Expression") +
  theme_minimal() +
  theme(legend.position = "none")
ggsave("figures/boxplot.svg", p, width = 6, height = 4)
cat("Saved figures/boxplot.svg\n")

# ── Volcano plot ───────────────────────────────────────────────────

de$neglog10p <- -log10(de$padj)
de$sig <- ifelse(de$padj < 0.05 & de$log2fc > 1, "Up",
           ifelse(de$padj < 0.05 & de$log2fc < -1, "Down", "NS"))

p <- ggplot(de, aes(x = log2fc, y = neglog10p, color = sig)) +
  geom_point(alpha = 0.7) +
  scale_color_manual(values = c("Up" = "red", "Down" = "blue", "NS" = "grey60")) +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed", color = "grey50") +
  geom_vline(xintercept = c(-1, 1), linetype = "dashed", color = "grey50") +
  labs(title = "A) Differential Expression",
       x = "log2 Fold Change", y = "-log10(adjusted p-value)") +
  theme_minimal()
ggsave("figures/fig1_volcano.svg", p, width = 8, height = 6)
cat("Saved figures/fig1_volcano.svg\n")

# ── MA plot ────────────────────────────────────────────────────────

de$log10mean <- log10(de$baseMean)
p <- ggplot(de, aes(x = log10mean, y = log2fc, color = sig)) +
  geom_point(alpha = 0.7) +
  scale_color_manual(values = c("Up" = "red", "Down" = "blue", "NS" = "grey60")) +
  geom_hline(yintercept = 0, linetype = "dashed", color = "grey50") +
  labs(title = "B) MA Plot",
       x = "log10(baseMean)", y = "log2 Fold Change") +
  theme_minimal()
ggsave("figures/fig2_ma.svg", p, width = 8, height = 6)
cat("Saved figures/fig2_ma.svg\n")

# ── Summary ────────────────────────────────────────────────────────

up <- sum(de$padj < 0.05 & de$log2fc > 1)
down <- sum(de$padj < 0.05 & de$log2fc < -1)
ns <- nrow(de) - up - down

cat(sprintf("\nUpregulated (padj<0.05, log2FC>1):   %d\n", up))
cat(sprintf("Downregulated (padj<0.05, log2FC<-1): %d\n", down))
cat(sprintf("Not significant:                      %d\n", ns))
cat("\nDone.\n")
